docs/extending-and-experimenting.md

```diff
@@ -147,3 +147,130 @@ subscribe to the result of `executeQuery`, keep track
+Since day one, with `urql`, we have aimed to make adaptability our number one priority through the use of _Exchanges_.
+Lets run through how this works by creating our own Dedup exchange - a middleware to debounce incoming query operations.
+
+> Note: We've already provided a bunch of exchanges that we thought you would find
+> useful. Check out the [exchanges source code](https://github.com/FormidableLabs/urql/blob/master/src/exchanges) for examples and usage information.
+
+Before you begin, it is assumed that you understand the fundamentals of _Operations_
+and _Exchanges_ in `urql`. There's a detailed description in the [Architecture](./architecture.md) section of the docs, but here's a quick recap:
+
+- Every operation in `urql` is passed to exchanges
+- Each exchange can do one of two things
+  - Forward the operation to the next exchange in line
+  - Return a result of the operation to the client
+
+First thing first, lets set up our boilerplate by making an _Exchange_ that takes an operation stream and passes it immediately to the next exchange in line. We'll also add in some initial logic which will later allow us to track which operations are in progress.
+
+```ts
+import { filter, pipe, tap } from 'wonka';
+import { Exchange } from 'urql';
+
+export const dedupExchange: Exchange = ({ forward }) => {
+  const inFlight = new Set<string>();
+
+  return ops$ => forward(ops$);
+};
+```
+
+You can see from the above example that an exchange is a [curried function](https://medium.com/javascript-scene/curry-and-function-composition-2c208d774983). It takes some setup arguments and returns a function which receives an operation stream (`ops$`) and forwards it to the next exchange (`forward`).
+
+Next thing we want to do is track unique incoming operations and add them to our tracking set. We also want to delete them from our set once the operation has completed. In order to do this we will be using the `pipe` and `tap` functions provided by `wonka`.
+
+> Note: The easiest way to uniquely identify an operation is by checking it's `key` property. This is a uniquely generated hash intended for this purpose.
+
+```ts
+import { filter, pipe, tap } from 'wonka';
+import { Exchange } from 'urql';
+
+export const dedupExchange: Exchange = ({ forward }) => {
+  const inFlight = new Set<string>();
+
+  return ops$ => {
+    const preFlight$ = pipe(
+      ops$,
+      // "tap" is called every time a new operation comes through.
+      // "tap" is invisible to subsequent stream operations
+      tap(operation => {
+        if (operation.operationName !== 'query') {
+          return;
+        }
+
+        inFlight.add(operation.key);
+      })
+    );
+
+    return pipe(
+      forward(preFlight$),
+      // This `tap` is called after an operation has been completed
+      tap(response => inFlight.delete(response.operation.key))
+    );
+  };
+};
+```
+
+Our exchange now tracks operations and their completion, but for this to be useful,
```

Comment:
I think at this point we'll need to explain the operators in detail and how the operations now flow through this exchange